ctapipe/image/time_next_neighbor_cleaning.py

```diff
@@ -0,0 +1,742 @@
+import numpy as np
+from scipy.spatial import cKDTree as KDTree
+from scipy.interpolate import interp1d
+
+import astropy.units as u
+import warnings
+import json
+
+
+def get_n_order_neighbor_matrix(geometry, order):
+    """
+    Get a matrix of the first and second order neighbors for each pixel.
+
+    Parameters
+    ----------
+    geometry
+
+    Returns
+    -------
+    Boolean matrix of dimension (n_pix, n_pix) with the first and second
+    neighbors for all pixels.
+    """
+    neighbors = geometry.neighbor_matrix_sparse
+    for i in range(order - 1):
+        neighbors = neighbors.dot(geometry.neighbor_matrix_sparse)
+    neighbors = neighbors.toarray()
+    np.fill_diagonal(neighbors, 0)
+    return neighbors
+
+
+def dilate_matrix(mask, matrix):
+    return np.dot(matrix, mask)
+
+
+def check_number_of_neighbors(mask, geo, min_neighbors=3, order=2, keep_neighbors=True):
+    neighbor_matrix = get_n_order_neighbor_matrix(geo, order)
+    n_neighbors = np.sum(mask * neighbor_matrix, axis=1) - 1
+
+    valid_pixels = n_neighbors >= min_neighbors
+    if keep_neighbors:
+        valid_pixels = dilate_matrix(valid_pixels, neighbor_matrix)
+
+    return valid_pixels & mask
+
+
+class TimeNextNeighborCleaning:
+    """
+    Class to perform the time next-neighbor cleaning defined in
+    https://arxiv.org/pdf/1307.4939.pdf . During the cleaning procedure, it is
+    searched for groups of pixels with a given charge and time-coincidence.
+
+    It requires only one parameter which is the fake image proability. The
+    cuts in the Q-dT parameterspace then are determined from the individual
+    pixel rate which gives the rate in which the pixels have a charge above
+    a given threshold. Those graphs are calculated from calibration data
+    (i.e. images without Cherenkov signal but pure night sky background).
+    """
+
+    def __init__(self):
+        self.IPR_dict = None
+        self._multiplicity_dict = {"4nn": 4, "3nn": 3, "2+1": 3, "2nn": 2}
+        self.__geometry = None
```

Comment:
generally you should not use double-underscores unless you really want name-mangling to occur (double underscores affect how  python interprets a member). A single-underscore is fine for showing that this is a protected member.  You could simply make the `geometry` property public to avoid the name collision. .

Reply:
All this hidden state in the class can go by properly making this a `TelescopeComponent` with a `subarray` and then giving a `tel_id` in `__call__`.